app/Main.hs

```diff
@@ -96,49 +96,108 @@ writePackageFile =
   . packageConfigToJSON
 
 data PackageInfo = PackageInfo
-  { repo         :: Text
+  { repo         :: Repo
   , version      :: Text
   , dependencies :: [PackageName]
   } deriving (Show, Eq, Generic, Aeson.FromJSON, Aeson.ToJSON)
 
 type PackageSet = Map.Map PackageName PackageInfo
 
+newtype Repo = Repo { unRepo :: Text } deriving (Show, Eq)
+
+instance Aeson.FromJSON Repo where
+  parseJSON = fmap Repo . Aeson.parseJSON
+
+
+instance Aeson.ToJSON Repo where
+  toJSON = Aeson.toJSON . unRepo
+
+
+data CloneTarget = CloneTag Text
+                 | CloneSHA Text
+                 deriving (Show)
+
+toCloneTarget
+  :: Repo
+  -> Text
+  -> IO CloneTarget
+toCloneTarget (Repo from) raw = do
+  remoteLines <- Turtle.fold (lineToText <$> gitProc) Foldl.list
+  let refs = Set.fromList (mapMaybe parseRef remoteLines)
+  if Set.member rawAsBranch refs
+     then do
+       echoT (raw <> " is a branch. psc-package only supports tags and SHAs.")
+       exit (ExitFailure 1)
+     else if Set.member rawAsTag refs
+             then return (CloneTag raw)
+             else return (CloneSHA raw)
+  where
+    rawAsBranch = "refs/heads/" <> raw
+    rawAsTag = "refs/tags/" <> raw
+    gitProc = inproc "git"
+      ["ls-remote"
+      , "-q"
+      , "--refs"
+      , "--heads"
+      , "--tags"
+      , from
+      ] empty
+    parseRef line = case T.splitOn "\t" line of
+      [_, ref] | "refs/" `T.isPrefixOf` ref -> Just ref
+      _ -> Nothing
+
+-- Both tags and SHAs can be treated as immutable so we only have to run this once
 cloneShallow
-  :: Text
+  :: Repo
   -- ^ repo
-  -> Text
-  -- ^ branch/tag
+  -> CloneTarget
+  -- ^ tag/SHA
   -> Turtle.FilePath
   -- ^ target directory
-  -> IO ExitCode
-cloneShallow from ref into =
-  proc "git"
-       [ "clone"
-       , "-q"
-       , "-c", "advice.detachedHead=false"
-       , "--depth", "1"
-       , "-b", ref
-       , from
-       , pathToTextUnsafe into
-       ] empty .||. exit (ExitFailure 1)
+  -> IO ()
+cloneShallow (Repo from) tgt into = do
+  void $ proc "git"
```

Comment:
When I try this with a SHA hash, I get `fatal: Remote branch <sha> not found in upstream origin`

Reply:
Did we confirm this works by now?